Components/Sensors/BatteryTask.cpp

```diff
@@ -0,0 +1,181 @@
+ */
+BatteryTask::BatteryTask() : Task(BATTERY_TASK_QUEUE_DEPTH_OBJS)
+{
+    data = (BatteryData*)soar_malloc(sizeof(BatteryData));
+}
+
+/**
+ * @brief Creates a task for the FreeRTOS Scheduler
+ */
+void BatteryTask::InitTask()
+{
+    // Make sure the task is not already initialized
+    SOAR_ASSERT(rtTaskHandle == nullptr, "Cannot initialize battery task twice");
+
+    // Start the task
+    BaseType_t rtValue =
+        xTaskCreate((TaskFunction_t)BatteryTask::RunTask,
+            (const char*)"BatTask",
+            (uint16_t)BATTERY_TASK_STACK_DEPTH_WORDS,
+            (void*)this,
+            (UBaseType_t)BATTERY_TASK_RTOS_PRIORITY,
+            (TaskHandle_t*)&rtTaskHandle);
+
+    //Ensure creation succeded
+    SOAR_ASSERT(rtValue == pdPASS, "BatteryTask::InitTask() - xTaskCreate() failed");
+}
+
+/**
+ * @brief BatteryTask run loop
+ * @param pvParams Currently unused task context
+ */
+void BatteryTask::Run(void * pvParams)
+{
+    while (1) {
+        Command cm;
+
+        //Wait forever for a command
+        qEvtQueue->ReceiveWait(cm);
+
+        //Process the command
+        HandleCommand(cm);
+    }
+}
+
+/**
+ * @brief Handles a command
+ * @param cm Command reference to handle
+ */
+void BatteryTask::HandleCommand(Command& cm)
+{
+    //Switch for the GLOBAL_COMMAND
+    switch (cm.GetCommand()) {
+    case REQUEST_COMMAND: {
+        HandleRequestCommand(cm.GetTaskCommand());
+    }
+    case TASK_SPECIFIC_COMMAND: {
+        break;
+    }
+    default:
+        SOAR_PRINT("BatteryTask - Received Unsupported Command {%d}\n", cm.GetCommand());
+        break;
+    }
+
+    //No matter what we happens, we must reset allocated data
+    cm.Reset();
+}
+
+/**
+ * @brief Handles a Request Command
+ * @param taskCommand The command to handle
+ */
+void BatteryTask::HandleRequestCommand(uint16_t taskCommand)
+{
+    //Switch for task specific command within REQUEST_COMMAND
+    switch (taskCommand) {
+    case BATTERY_REQUEST_NEW_SAMPLE:
+        SampleBatteryVoltage();
+        break;
+    case BATTERY_REQUEST_TRANSMIT:
+    	TransmitProtocolBatteryData();
+        break;
+    case BATTERY_REQUEST_DEBUG:
+        SOAR_PRINT("|VOLTAGE_TASK| Battery Voltage (V): %d.%d, MCU Timestamp: %u\r\n", data->voltage_ / 1000, data->voltage_ % 1000,
+        timestampPT);
+        SOAR_PRINT("Battery Power State: %d, \r\n", GetPowerState());
+        break;
+    default:
+        SOAR_PRINT("BATTERYTask - Received Unsupported REQUEST_COMMAND {%d}\n", taskCommand);
+        break;
+    }
+}
+
+/**
+ * @brief This function reads and updates battery voltage reading
+ *          from the battery.
+ */
+void BatteryTask::SampleBatteryVoltage()
+{
+	static const int BATTERY_VOLTAGE_ADC_POLL_TIMEOUT = 50;
+	double adcVal[1] = {};
+	double batteryVoltageValue = 0;
+	double vi = 0;
+	constexpr double adcConversion = 3.3/4095;
+	constexpr int voltageDividerScale = 4; // Value to scale voltage back to original value
```

Comment:
nit: put at the top of the .cpp under the Macros section so we can adjust easily